### PR TITLE
Document convert_concl_no_check (#3225)

### PR DESCRIPTION
**Pull request by Blaisorblade**
Triggered by trying to understand https://gitlab.mpi-sws.org/iris/iris/merge_requests/235. Probably not the priority, but I went to read Coq's source and almost documented its use instead. I tried to discourage beginners users from reading this by marking it as "advanced".

**Kind:** documentation.

### CODE REVIEW

File: doc/sphinx/proof-engine/tactics.rst
Comment (Zimmi48) [inline doc/sphinx/proof-engine/tactics.rst]:
```suggestion
      ensure safety. Hence, using :tacn:`change` checks convertibility twice
```

File: doc/sphinx/proof-engine/tactics.rst
Comment (Zimmi48) [inline doc/sphinx/proof-engine/tactics.rst]: You should use `:tacn:` here as well, but you need first to add a `:name:` option to the `.. tacv::` directive, otherwise it won't be indexed (because variants aren't by default).

**Review by Zimmi48 — COMMENTED** [review]
This is great :rocket: Thank you!

**Comment by herbelin** [comment]
Actually, what about directly providing a `change_no_check` tactic? That would then follow the same pattern as in `exact`/`exact_no_check`.

**Comment by Blaisorblade** [comment]
`change_no_check` seems indeed a more consistent name. Are the proposed semantics the same?

If so, we could use the same docs, and list `convert_concl_no_check` as a (deprecated?) variant, since the latter would still be used for a while.

I've also added docs for the other `_no_check` tactics, tho I'm much less happy with those docs, and I could move those to a separate PR.
IMHO they all belong together, but quite possibly in a separate section. I was also surprised the `_cast_no_check` tactics are variants of `exact`, not change.
And documenting them without introducing <: and <<: is awkward.

I'm also not sure about the semantics of casts in the kernel. Docs for `vm_compute` and `native_compute` say they "evaluate the goal". If `t: T1`, in goal `T2`, how does one typecheck `t <: T3`/ `t <<: T3`? It seems only `T3` is normalized specially, and then convertibility checks proceed normally between `T3`'s normal form and `T1`, and `T2` and `T3`; the expectation, I guess, is that those convertibility checks are then trivial because the involved terms are close.
In fact, it seems all tactics pick `T3 = T2`, tho I can build terms by hand that violate this.

**Comment by Blaisorblade** [comment]
BTW, are the failures on Nix anyhow due to my doc changes? I'll assume not. https://gitlab.com/coq/coq/-/jobs/203628478

File: doc/sphinx/proof-engine/tactics.rst
Comment (Zimmi48) [inline doc/sphinx/proof-engine/tactics.rst]: The location of this documentation doesn't seem right. If you want to have all the `no_check` tactic be documented close together, then you could create a new section for these tactics. But here, you chose to document them in the section on reduction tactics, which only makes sense for `convert_concl_no_check`.

**Comment by Zimmi48** [comment]
Related: #7846.

@herbelin Such uniformization of naming would make sense, but I suggest doing it in a separate PR to keep concerns separate.

**Comment by Zimmi48** [comment]
> BTW, are the failures on Nix anyhow due to my doc changes? I'll assume not. https://gitlab.com/coq/coq/-/jobs/203628478

Nope, these are test files from the test-suite that fail from time to time.

**Comment by JasonGross** [comment]
> I was also surprised the `_cast_no_check` tactics are variants of `exact`, not `change`.

Why is this a surprise?  `change` does not update the proof term, and takes a type, while, say, `vm_cast_no_check x` is roughly the same as `exact (x <: _)` (modulo checking); it takes a *term* and uses that the finish the proof.

**Comment by Zimmi48** [comment]
@Blaisorblade #10017 has been merged, so this PR could be updated to document it (and mark `convert_concl_no_check` as a deprecated variant).

**Comment by Blaisorblade** [comment]
I'll do that and resplit sections as suggested — I'm adding a new section at the end. I see now `change_no_check` gets `change`'s variants; that's good.

> Why is this a surprise? `change` does not update the proof term, and takes a type, while, say, `vm_cast_no_check x` is roughly the same as `exact (x <: _)` (modulo checking); it takes a _term_ and uses that the finish the proof.

To me, the obvious and orthogonal/compositional thing would have been an (oddly-named) version of `{vm/native}_compute_nocheck` to just insert a cast for the kernel without doing the computation; `vm_cast_no_check` can then be built on top, instead of being such an ad-hoc primitive combination.

**Comment by Blaisorblade** [comment]
Updated as planned — except I've split out docs for change_no_check's variants in #10053 — see there.

File: doc/sphinx/proof-engine/tactics.rst
Comment (JasonGross) [inline doc/sphinx/proof-engine/tactics.rst]:
```suggestion
   overall, while :tacn:`change_no_check` can produce ill-typed terms,
```

File: doc/sphinx/proof-engine/tactics.rst
Comment (JasonGross) [inline doc/sphinx/proof-engine/tactics.rst]:
```suggestion
   Hence, :tacn:`change_no_check` can be useful to speed up certain proof
```

File: doc/sphinx/proof-engine/tactics.rst
Comment (JasonGross) [inline doc/sphinx/proof-engine/tactics.rst]:
```suggestion
   In the following example, :tacn:`change_no_check` replaces :g:`False` by
```

File: doc/sphinx/proof-engine/tactics.rst
Comment (JasonGross) [inline doc/sphinx/proof-engine/tactics.rst]:
```suggestion
   check instead. See :tacn:`change_no_check` for more explanations.
```

**Review by JasonGross — CHANGES_REQUESTED** [review]

File: doc/sphinx/proof-engine/tactics.rst
Comment (JasonGross) [inline doc/sphinx/proof-engine/tactics.rst]: Minor grammar nit (not strictly necessary, but I think it reads better):
```suggestion
   it skips checking that :n:`@term` is convertible with the goal.
```

File: doc/sphinx/proof-engine/tactics.rst
Comment (JasonGross) [inline doc/sphinx/proof-engine/tactics.rst]: Minor grammar nit:
```suggestion
   Recall that the Coq kernel typechecks proofs again when they are concluded to
```

File: doc/sphinx/proof-engine/tactics.rst
Comment (Blaisorblade) [reply doc/sphinx/proof-engine/tactics.rst]: Your call, but I wouldn't. The manual prefers "A and B are convertible", followed by "convertible to", then "convertible with".

**Review by Zimmi48 — APPROVED** [review]

**Comment by Zimmi48** [comment]
Waiting on @JasonGross's approval.

**Review by JasonGross — APPROVED** [review]

**Pull request merged**
